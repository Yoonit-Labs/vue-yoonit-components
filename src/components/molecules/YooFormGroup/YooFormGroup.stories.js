import yooFormGroup from './YooFormGroup.vue'

export default {
  title: 'Yoonit/YooFormGroup',
  component: yooFormGroup
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooFormGroup },
  template:
    '<yoo-form-group v-bind="$props"></yoo-form-group>'
})

export const Custom = Template.bind({})
Custom.args = { label: 'Label', placeholder: 'Placeholder', validateOnDataInput: true }

export const Disabled = Template.bind({})
Disabled.args = { label: 'Label', placeholder: 'Placeholder', disabled: true }
