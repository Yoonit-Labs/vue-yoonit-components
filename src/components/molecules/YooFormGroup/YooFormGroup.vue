<template lang="pug">
  div(
    :class="['yoo-input-group']"
  )
    label(
      :class="['yoo-input', ...takeModifier]"
    )
      input(
        :class="['yoo-input__field']"
        :value="value"
        v-on="listeners"
        v-bind="$attrs"
        @input="handleInput"
        @blur="doValidate"
        ref="textField"
      )

      span(
        class="yoo-input__label"
      ) {{ label }}
        i.fa

      span(
        :class="['yoo-input__feedback']",
        v-show="!isValid"
      ) {{ invalidMessage }}

</template>

<script>

export default {
  name: 'YooFormGroup',
  props: {
    label: {
      type: String
    },
    value: {
      type: String
    },
    required: {
      type: Boolean
    },
    disabled: {
      type: Boolean
    },
    requiredRule: {
      type: Function
    },
    validateOnDataInput: {
      type: Boolean,
      default: false
    }
  },
  components: {},
  data: () => ({
    isValid: '',
    invalidMessage: ''
  }),
  beforeCreate () {},
  created () {},
  beforeMount () {},
  mounted () {
    if (this.disabled) this.$refs.textField.disabled = true

    if (this.required) {
      this.value === ''
        ? this.isValid = false
        : this.isValid = true
    }
  },
  beforeUpdate () {},
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  computed: {
    listeners () {
      const {
        input,
        ...others
      } = this.$listeners
      // protect custom input
      return { ...others }
    },
    takeModifier () {
      const block = 'yoo-input'
      const classList = []

      if (this.isValid === false) {
        classList.push(`${block}--error`)
      }

      if (this.isValid === true) {
        classList.push(`${block}--success`)
      }

      return classList
    }
  },
  methods: {
    async handleInput (event) {
      await this.$emit('input', event.target.value)
      if (this.validateOnDataInput) {
        this.doValidate(event)
      }
    },
    doValidate (e) {
      if (this.required) {
        this.value === ''
          ? this.isValid = false
          : this.isValid = true

        this.invalidMessage = `O campo ${this.label} é obrigatório.`
      }

      if (this.value && this.requiredRule) {
        const {
          result,
          message
        } = this
          .requiredRule(this.value)

        this.isValid = result
        this.invalidMessage = message
      }
      this.$emit('update:isValid', this.isValid)
    }
  },
  filters: {},
  watch: {}
}
</script>

<style src="./YooFormGroup.sass" lang="sass" scoped></style>
