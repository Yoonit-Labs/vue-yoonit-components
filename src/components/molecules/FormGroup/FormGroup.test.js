import { shallowMount } from '@vue/test-utils'
import YooFormGroup from '@/components/molecules/FormGroup/FormGroup.vue'
const mountComponent = () => {
  return shallowMount(YooFormGroup, {
    propsData: {
      required: true,
      label: 'Label',
      value: 'Text',
      placeholder: 'Placeholder'
    },
    data: () => ({
      userHasTyped: true
    })
  })
}
describe('YooFormGroup Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountComponent()
  })
  it('Matches Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('Loads the Component HTML', () => {
    expect(wrapper.find('input').exists()).toBe(true)
  })
  describe('Props', () => {
    describe('Disabled mode', () => {
      it('Check if this input is disabled', async () => {
        const wrapper = shallowMount(YooFormGroup, {
          propsData: {
            disabled: true,
            label: 'Disabled',
            placeholder: 'Placeholder'
          }
        })
        expect(wrapper.find('input').element.disabled).toBeTruthy()
      })
    })
    describe('isValid Test', () => {
      it('If not isValid state', async () => {
        await wrapper.setProps({ isValid: false })
        expect(wrapper.find('label').classes().includes('yoo-input--error')).toBe(true)
      })
      it('If isValid state', async () => {
        await wrapper.setProps({ isValid: true })
        expect(wrapper.find('label').classes().includes('yoo-input--success')).toBe(true)
      })
    })
    describe('Required Test', () => {
      it('If required and text is empty', async () => {
        await wrapper.setProps({ validateOnDataInput: false, value: '', required: true, isValid: false })
        await wrapper.find('input').trigger('blur')
        expect(wrapper.find('label').classes().includes('yoo-input--error')).toBe(true)
      })
      it('If required false', async () => {
        await wrapper.setProps({ validateOnDataInput: false, value: '', required: false, isValid: true })
        await wrapper.find('input').trigger('blur')
        expect(wrapper.find('label').classes().includes('yoo-input--error')).toBe(false)
      })
      it('If required and has text', async () => {
        await wrapper.setProps({ value: 'Any text', isValid: true })
        expect(wrapper.vm.required).toBeTruthy()
        await wrapper.find('input').trigger('blur')
        expect(wrapper.find('label').classes().includes('yoo-input--success')).toBe(true)
      })
    })
    describe('Required Rule', () => {
      it('Required rule is executed', async () => {
        await wrapper.setProps({
          validateOnDataInput: true,
          value: 'a',
          requiredRule: value => {
            let status = true
            if (value.length < 3) {
              status = false
            }
            return {
              result: status,
              message: 'The field Text must be more then 3 characters.'
            }
          }
        })
        await wrapper.find('input').setValue('a')
        expect(wrapper.find('label').classes().includes('yoo-input--error')).toBe(true)
      })
    })
  }) // describe Props
  describe('Events', () => {
    describe('On input text', () => {
      it('Emits keypress Event', async () => {
        await wrapper.setProps({ validateOnDataInput: false })
        await wrapper.find('input').setValue('a')
        expect(wrapper.emitted()).toHaveProperty('input')
      })
    })
  }) // describe Events
})
