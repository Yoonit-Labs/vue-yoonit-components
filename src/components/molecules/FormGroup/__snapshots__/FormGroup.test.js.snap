// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooFormGroup Component Matches Snapshot 1`] = `<div class="yoo-input-group" placeholder="Placeholder"><label class="yoo-input yoo-input--success"><input placeholder="Placeholder" class="yoo-input__field"><span class="yoo-input__label">Label<i class="fa"></i></span><span class="yoo-input__feedback" style="display: none;"></span></label></div>`;
